sample/net/dummy/connector/replica-connector.go

```diff
@@ -19,39 +19,35 @@
 package connector
 
 import (
-	"fmt"
-
 	"github.com/hyperledger-labs/minbft/api"
 )
 
 // ReplicaConnector allows to connect replica instances in the same
 // process directly. It is useful for testing.
 type ReplicaConnector struct {
-	replicas map[uint32]api.MessageStreamHandler
+	replicas map[uint32]chan api.MessageStreamHandler
 }
 
 var _ api.ReplicaConnector = (*ReplicaConnector)(nil)
 
 // New creates an instance of ReplicaConnector
-func New() *ReplicaConnector {
-	return &ReplicaConnector{
-		replicas: make(map[uint32]api.MessageStreamHandler),
+func New(n int) *ReplicaConnector {
+	replicas := make(map[uint32]chan api.MessageStreamHandler)
+	for i := 0; i < n; i++ {
+		replicas[uint32(i)] = make(chan api.MessageStreamHandler)
 	}
+	return &ReplicaConnector{replicas: replicas}
 }
 
 // ReplicaMessageStreamHandler returns the instance previously
 // assigned to replica ID.
 func (c *ReplicaConnector) ReplicaMessageStreamHandler(replicaID uint32) (api.MessageStreamHandler, error) {
-	replica, ok := c.replicas[replicaID]
-	if !ok {
-		return nil, fmt.Errorf("No message stream handler registered for replica %d", replicaID)
-	}
-	return replica, nil
+	return &messageStreamHandler{replicaID, c}, nil
 }
 
 // ConnectReplica assigns a replica instance to replica ID.
 func (c *ReplicaConnector) ConnectReplica(replicaID uint32, replica api.MessageStreamHandler) {
-	c.replicas[replicaID] = replica
+	c.replicas[replicaID] <- replica
```

Comment:
This line will only wake a single goroutine created by `messageStreamHandler.HandleMessageStream` and waiting on the channel (or block, if there is no one), whereas the API doesn't forbid calling `MessageStreamHandler.HandleMessageStream` multiple times on the same instance. The replica representation should behave in this case as if it was a real replica instance: establish multiple independent connections. 

Closing a channel, on the other hand, is a non-blocking broadcast signal to all goroutines waiting in receive operation on that channel. This issue should be easy to solve by using pre-allocated slices of replica instances and ready channels. Please look at https://github.com/hyperledger-labs/minbft/pull/70#discussion_r263558017.